httpclient/src/main/kotlin/tbdex/sdk/httpclient/Utils.kt

```diff
@@ -33,35 +37,89 @@ fun getPfiServiceEndpoint(pfiDid: String): String {
 /**
  * Generate request token.
  *
- * @param did
+ * @param did DID of the token creator
+ * @param pfiDid DID of the PFI
  * @param assertionMethodId
- * @return
+ * @return signed request token to be included as Authorization header for sending to PFI endpoints
+ *
  */
-fun generateRequestToken(did: Did, assertionMethodId: String? = null): String {
-  val assertionMethod = CryptoUtils.getAssertionMethod(did, assertionMethodId)
+fun generateRequestToken(did: Did, pfiDid: String, assertionMethodId: String? = null): String {
+
+  val didResolutionResult = DidResolvers.resolve(did.uri)
+  val assertionMethod: VerificationMethod = didResolutionResult.didDocument.findAssertionMethodById(assertionMethodId)
 
   // TODO: ensure that publicKeyJwk is not null
   val publicKeyJwk = JWK.parse(assertionMethod.publicKeyJwk)
   val keyAlias = did.keyManager.getDeterministicAlias(publicKeyJwk)
 
+  // TODO: figure out how to make more reliable since algorithm is technically not a required property of a JWK
   val algorithm = publicKeyJwk.algorithm
   val jwsAlgorithm = JWSAlgorithm.parse(algorithm.toString())
 
-  val jwsHeader = JWSHeader.Builder(jwsAlgorithm)
-    .keyID(assertionMethod.id.toString())
-    .build()
+  val kid = when (assertionMethod.id.isAbsolute) {
+    true -> assertionMethod.id.toString()
+    false -> "${did.uri}${assertionMethod.id}"
+  }
 
-  val payload = mapOf("timestamp" to Instant.now().toString())
-  val jwsPayload = Payload(payload)
-  val base64UrlEncodedPayload = jwsPayload.toBase64URL().toString()
+  val jwtHeader = JWSHeader.Builder(jwsAlgorithm)
+    .type(JOSEObjectType.JWT)
+    .keyID(kid)
+    .build()
 
-  val jwsObject = JWSObject(jwsHeader, jwsPayload)
-  val toSign = jwsObject.signingInput
+  val now = Instant.now()
+  val exp = now.plusSeconds(60)
+  val jwtPayload = JWTClaimsSet.Builder()
+    .audience(pfiDid)
+    .issuer(did.uri)
+    .expirationTime(Date.from(exp))
+    .issueTime(Date.from(now))
+    .jwtID(UuidCreator.getTimeOrderedEpoch().toString())
+    .build()
 
-  val signedBytes = did.keyManager.sign(keyAlias, toSign)
-  val base64UrlEncodedSignature = Base64URL(Convert(signedBytes).toBase64Url(padding = false))
-  val base64UrlEncodedHeader = jwsHeader.toBase64URL()
+  val jwtObject = SignedJWT(jwtHeader, jwtPayload)
+  val toSign = jwtObject.signingInput
+  val signatureBytes = did.keyManager.sign(keyAlias, toSign)
 
+  val base64UrlEncodedHeader = jwtHeader.toBase64URL()
+  val base64UrlEncodedPayload = jwtPayload.toPayload().toBase64URL()
+  val base64UrlEncodedSignature = Base64URL(Convert(signatureBytes).toBase64Url(padding = false))
 
   return "$base64UrlEncodedHeader.$base64UrlEncodedPayload.$base64UrlEncodedSignature"
+}
+
+fun verifyRequestToken(token: String, pfiDid: String): String {
+  val jwt: SignedJWT
+  try {
+    jwt = SignedJWT.parse(token)
+    // todo: resolving header.kid against a didresolver
```

Comment:
put these todos in a new issue: https://github.com/TBD54566975/tbdex-kt/issues/128